### DISCUSSION TITLE
How to run multiple self-hosted runners on a single host

### DISCUSSION THREAD

**Discussion by dmerrick** [community/community#26258] [Actions]
<p>We’re using self-hosted runners in our organization and it’s working fine. What I want to do now is use bigger instances and run multiple runners per host.</p>
<p>I have a script that uses a PAT to generate a runner token, and then uses that token to start a runner. When I run the script multiple times on the same host, only one will register correctly. I get errors like:</p>
<pre><code class="lang-auto">A session for this runner already exists.
2020-09-03 19:26:14Z: Runner connect error: The actions runner 94e9c2e0ca90995c7764c845 already has an active session for owner ip-20-10-43-71.. Retrying until reconnected.
</code></pre>
<p>or</p>
<pre><code class="lang-auto">Cannot configure the runner because it is already configured. To reconfigure the runner, run 'config.cmd remove' or './config.sh remove' first.
</code></pre>
<p>I have noticed that the token I get from the <a href="https://docs.github.com/en/rest/reference/actions#create-a-registration-token-for-an-organization" rel="nofollow noopener">token registration endpoint</a> is the same for each runner I try to launch. While this might be the issue, I see no way to “force” the generation of a new registration token.</p>
<p>I should also mention using a different (randomly-generated) <code>--name</code> and <code>--workdir</code> when I run <code>config.sh</code>.</p>

**Comment by weide-zhou** [comment, marked as answer]
<p>Hi <a class="mention" href="/dmerrick">@dmerrick</a>,</p>
<p>Glad to see you in Github Community Forum!</p>
<p>The error should be due to you didn’t change the directory for the configuration script(config.sh/config.cmd), please change to another workdir for new configuration.</p>
<p>To install multiple self-hosted runner on your host, you should use different workdir and different runner name. The registration token can be identical( or you can rerun the <a href="https://docs.github.com/en/rest/reference/actions#create-a-registration-token-for-an-organization" rel="nofollow noopener">api</a> command to get a new one).</p>
<p>These runners are using same machine env and disk, it’s recommened to create them on different VMs on your host to seperate them.</p>
<p>Thanks</p>

**Reply by kluevandrew** [reply]
Any option to not copy runner binaries? If i've change working directory and make symlinks to binaries and shell scripts it's doesn't work.

**Reply by akvadrako** [reply]
I also couldn't find a way to run them with a shared release directory and I suppose that would get complex with the self-updating feature. Though if you handle updates yourself and use BTRFS / XFS you could use reflinks instead of copies, so they don't take more space.

**Comment by dmerrick** [comment]
<p>Thank you, I was able to get this working by running <code>config.sh</code> from different directories for each worker.</p>

**Reply by Timmmm** [reply]
This didn't work for me. It checks for `./bin/libcoreclr.so` and other files which only exist if you create full copies of the runner, rather than just running `config.sh` from different directories.

**Comment by PradeepKiruvale** [comment]
<p>Hi Dmerrick,</p>
<p>I tried to run multiple runners on my self-hosted runners as well. I am getting the below error. Even then I used different working directories for different runners. Could you please let me know what mistake I am doing?</p>
<p>Cannot configure the runner because it is already configured. To reconfigure the runner, run ‘config.cmd remove’ or ‘./config.sh remove’ first.</p>

**Comment by rahul-kumar-saini** [comment]
<p>Hey Pradeep,</p>
<p>I’ve got a setup working for starting multiple self-hosted runners on the same host, for different repositories.</p>
<p>I have a directory for each repository, and extract the actions runner .tar into each one separately. I generate a token and then run the config.sh inside each directory.</p>
<p>So my file structure looks something like:<br>
base/<br>
|-dir1<br>
|-|-config.sh<br>
|-dir2<br>
|-|-config.sh<br>
…</p>
<p>If you’ve got a similar setup, then one thing you may want to check is whether or not you’re using the “–name” positional argument for config.sh. You need to make sure each one has a unique name.</p>
<p>Eg:</p>
<p>./config.sh --name &lt;unique_name&gt; --url …</p>
<p>Hope this helps!</p>

**Comment by PradeepKiruvale** [comment]
<p>Thanks for the reply Rahul. I wanted two runners for the same repo, is it possible?</p>

**Reply by Deepak17460** [reply]
Hi @PradeepKiruvale , Yes it is possible!

Just follow these steps:-

**How to Label Runners**
Navigate to Repository Settings:

Go to your GitHub repository.
Click on Settings.
Navigate to Actions > Runners.
Add Labels to Runners:

You will see a list of your self-hosted runners.
Click on the runner you want to label (e.g., dpcode).
Click on Edit to add or modify labels.
Add the label dpcode for the runner that should handle most jobs.
Add the label dpcode-1 for the runner that should handle the SonarQube analysis job.

**Comment by rahul-kumar-saini** [comment]
<p>Yep it is possible, I’ve tested it.</p>
<p>Just make sure you run config.sh out of separate directories, and make the runner names unique!</p>

**Comment by PradeepKiruvale** [comment]
<p>OK, thanks Rahul, I will have a look.</p>

**Comment by blugee** [comment]
<p>@ <a href="https://github.community/u/rahul-kumar-saini">rahul-kumar-saini</a> can you please share some examples of how to deploy multiple runners on a single host for single repository different branches.</p>
<p>My scenario is like, I have an ubuntu server where I want to access the develop branch and production branch separately.</p>
<p>Thanks</p>

**Comment by stardustman** [comment]
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/3X/2/3/2337595d3dbd98c6d40dd4f197a8785486c54eb4.png" data-download-href="/uploads/short-url/51xin3ub1m6B7lmYZQ3Y5MAewpm.png?dl=1" title="github-runner" rel="noopener nofollow ugc"><img src="https://user-images.githubusercontent.com/22998204/181096486-3b5ca322-5aff-4ef4-a68a-60140b34a2e4.png" alt="github-runner" data-base62-sha1="51xin3ub1m6B7lmYZQ3Y5MAewpm" width="690" height="174" srcset="https://aws1.discourse-cdn.com/github/optimized/3X/2/3/2337595d3dbd98c6d40dd4f197a8785486c54eb4_2_690x174.png, https://aws1.discourse-cdn.com/github/optimized/3X/2/3/2337595d3dbd98c6d40dd4f197a8785486c54eb4_2_1035x261.png 1.5x, https://aws1.discourse-cdn.com/github/optimized/3X/2/3/2337595d3dbd98c6d40dd4f197a8785486c54eb4_2_1380x348.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/3X/2/3/2337595d3dbd98c6d40dd4f197a8785486c54eb4_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">github-runner</span><span class="informations">3156×796 256 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>I have launched a github runner at <code>/root/actions-runner</code> directory, when I launch another at /runner/actions-runner on the same host, I have the same problem. Is there something wrong?</p>

**Reply by raphaelcangucu** [reply]
Probably you copied same folder,  try to remove the .runner file and try again.

**Comment by chiranjib-iag** [comment]
<p>Is this possible single git hub repo with multiple self hosted runner.under that repo one directory point to one runner another directory another runner.how can we do structure in entrypoint.sh if you could help.</p>

**Comment by praveenmak** [comment]
<p>Once we start the github runner for same repo, same host. How do we use that in our Github Actions?<br>
Please provide an example on how to specify in the “runs-on”</p>

**Comment by dgholz** [comment]
<aside class="quote no-group" data-username="praveenmak" data-post="12" data-topic="130474">
<div class="title">
<div class="quote-controls"></div>
 praveenmak:</div>
<blockquote>
<p>Please provide an example on how to specify in the “runs-on”</p>
</blockquote>
</aside>
<pre><code class="lang-auto">jobs:
  my-self-hosted-job:
    runs-on: self-hosted
</code></pre>
<p>See <a href="https://docs.github.com/en/actions/hosting-your-own-runners/using-self-hosted-runners-in-a-workflow" rel="noopener nofollow ugc">the docs about running on self-hosted runners</a> for more info. We usually <a href="https://docs.github.com/en/actions/hosting-your-own-runners/using-labels-with-self-hosted-runners#using-the-configuration-script-to-create-and-assign-labels" rel="noopener nofollow ugc">add a few more labels when registering the runner</a> so we can select e.g. by instance size.</p>

**Comment by ghost** [comment]
<p>how many self-hosted runners we can create in same host ? any limit ?</p>

**Comment by brycesnyder** [comment]
Lots of good information here, but I feel like it hasn't really been answered. 

Is this a way to have multiple runners from a single host? 

i.e I set up a runner and then link all calls to that runner and it just synchronously runs the calls?

Or, is there a way for us to simply have a local repo and when I run a command it fires up each instance required?

**Reply by GuillaumeBourque-QC** [reply]
Hello all,

you can have as many runners as you want.   The only limit is your machine capacity.

To make sure you get it running you must follow simple rules.

1.  if you run on a server with selinux enable dont work in a user home, go to a dir like /opt/github
2. all runner must run in a separate dir.
3. you must configure all the runner one by one.
4. the registration token can be identical, but the runner name must be unique with in your github repos.

Here is an example ( as of tody the runner version is 2.299.1, so replace accordingly.

## Preperation

This is done on an AWS ec2 vm with user ec2-user is is bettre tu run your runner under a user that is not root.

```
mkdir /opt/github/action-runner-1 
mkdir /opt/github/action-runner-2

cd opt/github/action-runner-1
curl -o actions-runner-linux-x64-2.299.1.tar.gz -L https://github.com/actions/runner/releases/download/v2.299.1/actions-runner-linux-x64-2.299.1.tar.gz
tar xzf ./actions-runner-linux-x64-2.299.1.tar.gz

# this will copy your runner source code from dir 1 to dir 2.  You must do this step before configuring your first runner.
rsync -av /opt/github/action-runner-1/ /opt/github/action-runner-2/

````

## Configuration

For runner # 1

```
cd opt/github/action-runner-1
./config.sh --url https://github.com/YourORG/YourRepo --token **** (Your token from github)

# You enter the group and a unique name: --> runner-1 

# here we  create a startup script on an AWS ec2 machine that as the current usermame  ec2-user 
sudo ./svc.sh install ec2-user 

# Then you start your new service depending on what you get from previous step
sudo systemctl start actions.runner.YourOrg-OrRepo.runner-1.service 

# You should see your firt runner in your github repos or Org

````

For runner # 2

```
cd opt/github/action-runner-2
./config.sh --url https://github.com/YourOrg/YourRepo --token **** (Your token from github)

# You enter the group and a unique name: --> runner-2

# here we  create a startup script on an AWS ec2 machine that as ec2-user 
sudo ./svc.sh install ec2-user 

#Then you start your new service depending on what you get from previous step
sudo systemctl start actions.runner.YourOrg-OrRepo.unner-2.service 

# You should see your second runner in your github repos or Org
```

Voilà it should work now with 2 runners.

**Reply by GuillaumeBourque-QC** [reply]
@ErcinDedeoglu I don't get your point, I responded to the many questions about "is it possible to run multiple selfhosted runner on a single host", Yes and I gave a detailed solution on how to do it manually.

As to the second question ""i.e I set up a runner and then link all calls to that runner and it just synchronously runs the calls?", since I have no clue I did not respond and the question is not clear .to me ....

**Reply by Tejalgnrml** [reply]
@GuillaumeBourque-QC 
Is the resource allocation for the multiple runners is dynamic based on the type of job we run?

**Reply by ggreco** [reply]
Following this logic I'm able to have two runners for two separate repos on the same machine, I see them "green" in the actions -> runners -> self-hosted panel, I see the labels correctly, no error log at all using journalctl, but only the first of the two is able to receive jobs, the second stays idle forever, it's possible that something changed in recent runner versions that make this workaround not possible anymore?

**Reply by sunnamed434** [reply]
But what if after doing these steps, want to add more runners in the future? Copy via rsync is ok again?!

**Comment by amhelderman** [comment]
I wanted to share a tip that might help those who want to generate 3+ runners programmatically but who don't want to have to grab new tokens. Although I hear those tokens can be shared, that might not be useful in an install script.

When using the `config.sh`, instead of using the `--token` option, you can use the `--pat` and `--unattended` options with `gh auth token` if you're logged into that github CLI.

Like so:
```
./config.sh --unattended \
                --name "$RUNNER_NAME_PREFIX"-"$i" \
                --url "$REPOSITORY_URL" \
                --pat $(gh auth token)
```

You can also use this to remove runners with `./config.sh remove --pat $(gh auth token)`.

(Please note that $RUNNER_NAME_PREFIX and $REPOSITORY_URL are specific to your needs and $i is just because I'm using this in a loop where $i is incrementing).

**Comment by vbem** [comment]
There is a [simple wrapper script](https://github.com/vbem/multi-runners) which setup multi self-hosted GitHub action runners on single host.

**Reply by gajus** [reply]
Thank you. This just saved me a lot of time!

**Comment by salimshamim** [comment]
nobody answered what do we put here 
`jobs:
  Build:
    name: Build
    runs-on: xxxxxx`

**Comment by abhi-iac** [comment]
I need to install the two self-hosted runners per repo for more than 100 teams.

Do anyone has a script which will install the runners onto the respective VMs and make the runners run as service? I need to install the script on Ubuntu and Win 2019 servers

**Reply by Tejalgnrml** [reply]
Hi you automate self-hosted runner creation using github api where you can get the registration token automatically. 
you can automate connecting to your VMs and create the runners.

**Comment by abhi-iac** [comment]
Thanks for your reply, do you have any script which does this process? I am
very new to GitHub.

Kind Regards,
Abhishek

**Comment by A7ryan** [comment]
![image](https://github.com/user-attachments/assets/023f861f-6036-48b1-9100-af69c11502ad)


________________________


### **SOLVED: How to run multiple self-hosted runners on a single host #26258**

Hello,
I have read the above answers and just wanted to add:
- You don’t need to install a separate runner for each repository!
- I have found something exciting.


________________________


- Follow Each Step in GitHub Organization (Guess what: it's all free)

Step 1: Just Create a free GitHub Organization

Step 2: Migrate/Copy your repos to you want to run to your organization.

Step 3: Create a Runner Group.

Step 4: Add the repos you want to this runner group.

Step 5: Create .yml for each repo (different files as per purpose)

Step 6: Create a new self-hosted Runner.

Step 7: Setup the runner on your windows/other OS.

Step 8: Provide the Runner Group and Name of runner (See the pic attached)

Step 9: Done (now all your repos will run on a single host via single runner)

Step 10: If you get a path specific error do check out my answer: https://github.com/actions/runner/issues/2540#issuecomment-3026724121

**Reply by Rod-in-NM** [reply]
@A7ryan I thought all GitHub Organizations needed to be paid for. If I'm wrong about that, then how does one get a free GitHub Organization?

**Reply by A7ryan** [reply]
https://github.com/user-attachments/assets/9598efbc-7749-44d4-9fbb-ee3f54c45450

_________________


Hello @Rod-in-NM,

You can create a free GitHub Organization. (I have attached the video it would guide you through)

- Open your Github Account (E.g. github.com/username)

- On Top Right (Click on your profile Pic)

- You will find an Option -> Your Organizations (click on that)

- Click on New Organization

- Click: Create a Free GitHub Organization

- Enter your details

- Congratulations, on creating your GitHub Organization!

- (If you face an issue related to captcha just change your browser/device.

- Let me know if you need any further assistance!

Best,
